### ISSUE TITLE
"Post" uniforms are not Post rendering

### ISSUE THREAD

**Issue by claudeha** [3Dickulus/FragM#56]
It would be very nice if the `Post` tab uniforms from the buffer shader could be modified without having to re-render every time.  Adjusting `Contrast` for example has no need for recalculation, all the data from the subframes has already been accumulated in the back buffer.

If this is implemented, it might be nice to have some visual indicator of which uniforms can be modified without recalculating, perhaps this could be some little green icon of some shape or other next to the lock?

But, to do this would need to have fixed the bug where uniforms present only in the buffer shader are not shown (and default to 0) unless they are duplicated into the main frag...

**Comment by 3Dickulus** [comment]
they (shaderprogram and buffershaderprogram) execute interleaved 1 to 1 unless the `#iterationsBetweenRedraws = n` directive is used, if this is set then buffershaderprogram executes once every n subframes.

currently no plans on changing the engine internal flow unless we're going to sit down and bash out a new modern engine that fits in place of the old one.

**Comment by claudeha** [comment]
sure, but setting uniforms only used in the buffer shader shouldn't clear accumulated subframes afaict

**Comment by claudeha** [comment]
Here's a minimal example showing the problem: the BufferShader.frag has a Post group with an Exposure uniform, but it isn't accessible through the GUI (there is only the Camera tab).

[Main.frag.txt](https://github.com/3Dickulus/FragM/files/3502895/Main.frag.txt)
[BufferShader.frag.txt](https://github.com/3Dickulus/FragM/files/3502896/BufferShader.frag.txt)

**Comment by 3Dickulus** [comment]
This is what my log shows...
```
Fragment shader compiled with warnings: 
Vertex info
-----------
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(4) : warning C7555: 'varying' is deprecated, use 'in/out' instead
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(5) : warning C7555: 'varying' is deprecated, use 'in/out' instead
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(15) : warning C7533: global variable gl_ProjectionMatrix is deprecated after version 120
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(15) : warning C7533: global variable gl_Vertex is deprecated after version 120

Fragment info
-------------
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(27) : warning C7555: 'varying' is deprecated, use 'in/out' instead
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(28) : warning C7555: 'varying' is deprecated, use 'in/out' instead
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(46) : warning C7533: global variable gl_FragColor is deprecated after version 120

Compiled script in 24 ms.
```

When I change #version to 120 I get this...
```
Could not create fragment shader: 
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(35) : error C7548: '^' requires "#extension GL_EXT_gpu_shader4 : enable" before use
/home/toonfish/Fragmentarium/claudes-test_Files/buffershadertest/Main.frag(35) : error C7548: '&' requires "#extension GL_EXT_gpu_shader4 : enable" before use

Failed to compile script (8 ms).
```

when I add
`#extension GL_EXT_gpu_shader4 : enable` to main.frag it compiles

ha, and the line numbers are all off, need "deeper" accounting of lines, the FragM specific tricks are turned into comments when the code is sent to compiler so they don't make it barf, and the vertex and fragment code is sent to the compiler separately then binaries are linked into a working program, I think this is one of those things that Syntopia abandoned as it starts to get very convoluted to keep track.

the buffershader uses a trick, in GLSL if different shaders have a uniform declared in the exact same way, name, type, value etc then the compiler references only one instance for both shaders so...

in main frag one would declare
```
#group Post
uniform float Exposure; slider[-100,0,100]
```
which is somewhat counter intuitive because Exposure is (must) not be used in the shaderfrag

and in buffershader.frag then declare
`uniform float Exposure; `

this makes it show up in the gui as a widget. using the above I get this...
![Screenshot_20190814_164819](https://user-images.githubusercontent.com/4978723/63063993-f75cfa80-beb3-11e9-87e6-c3e8a0b2753c.png)

**Comment by 3Dickulus** [comment]
to continue...
the uniform in the shaderfrag gets optimized out of the shaderbinary as it's not used but the widget remains because FragM finds the second uniform in the buffershader where it is used so doesn't get optimized out.
This is how I understand it, for a buffershader uniform to show as a widget it needs to be declared as a widget in the primary shader and declared in the secondary shader as a uniform only.

**Comment by claudeha** [comment]
Yes that's how it works now.  I'm saying that it's a bit wrong.  The buffer shader frag should have its own uniforms in the GUI, without also needing to declare them in the main frag.  And moreover, setting buffer shader uniforms from the GUI should not clear accumulated subframes.

Remains to consider what should happen if both shaders declare the same uniforms with different types.

**Comment by 3Dickulus** [comment]
yes, I agree, they should not be intertwined like that, they should be separate entities, this is something to work towards.

**Comment by claudeha** [comment]
I have something working for this.  I basically created another `VariableEditor` to hold the `BufferShader` stuff (now there are two, someone better at Qt than me might tweak the layout), and did some hackery so that changing BufferShader uniforms doesn't clear the accumulated subframes.

Screenshot: ![2019-08-18_fragm_buffershader_post](https://user-images.githubusercontent.com/10728495/63226679-b119db00-c1d4-11e9-8f49-6559fbbf7f28.png)

Bugs still to figure out:
- Subframe counter is reset when changing `BufferShader` widgets, this is misleading because the accumulated subframes are not lost.
- Works in Progressive mode with play enabled.  With play disabled, no effect from changing BufferShader uniforms.  Not sure why yet.
- Frags need modifying to put the GUI specification into the BufferShader where it belongs, otherwise they are deleted from the Main VariableEditor because they aren't used in the Main frag, and they aren't included in the BufferShader VariableEditor because they are not defined there.  This probably **breaks loading older saved frags** (which use the hacky workaround putting Post GUI specs in the Main frag even though they are unused there), so this solution may be something to postpone for FragM 2.6 or 3.0?

**Comment by claudeha** [comment]
I guess resetting subframe count will break Progressive2D etc rendering that uses subframe index for AA.

**Comment by 3Dickulus** [comment]
can it be backward compatible? so the current frag base still works?

**Comment by claudeha** [comment]
Maybe, it would be highly desirable indeed.  Would need to add back code to detect uniforms that are GUI-specified in Main but used only in BufferShader, and still display a GUI for them in the Main GUI (I don't think it would be feasible to move their GUI to the BufferShader variable editor, though I will look into it).  Possibly emit a WARNING in this case, so people can update their frags to put the GUI-spec into the BufferShader if they want to have post-rendering behaviour.

**Comment by claudeha** [comment]
Ok, I have the backward compatibility working I think, opening the Mandelbulb frag now warns:

```
Widget for ToneMapping should be specified in buffer shader!
Widget for Exposure should be specified in buffer shader!
Widget for Brightness should be specified in buffer shader!
Widget for Contrast should be specified in buffer shader!
Widget for AvgLumin should be specified in buffer shader!
Widget for Saturation should be specified in buffer shader!
Widget for LumCoeff should be specified in buffer shader!
Widget for Hue should be specified in buffer shader!
```

These widgets are displayed in the Main shader variable editor, and reset/lose accumulated subframes rendering when modified.

**Comment by claudeha** [comment]
Sorry for the rebase + force-push noise above.  Pull request here: <https://github.com/3Dickulus/FragM/pull/61>

**Comment by claudeha** [comment]
Going through all the frags to update them for the PR, found a problem with presets.  Presets are stored in the Main VariableEditor, and are not duplicated to the BufferShader VariableEditor.  Setting a preset using the GUI Apply button next to the preset list dropdown (which is only populated for the Main one) spits out warnings like this:

```
Could not find: Brightness
Could not find: Contrast
Could not find: Exposure
Could not find: Saturation
Could not find: ToneMapping
```

and indeed those BufferShader widgets are not updated.

Will try to figure out a way to solve this, possibly involving moving the preset manager out of the variable editor.  So don't merge the PR yet!

**Comment by claudeha** [comment]
Another bug: if there is no BufferShader at all, the widgets from the previous BufferShader are still shown.

**Comment by claudeha** [comment]
> Another bug: if there is no BufferShader at all, the widgets from the previous BufferShader are still shown.

Fixed this partially, but there is an empty Post tab instead of no tabs at all.

**Comment by claudeha** [comment]
> Bugs still to figure out:
> 
>     * Subframe counter is reset when changing `BufferShader` widgets, this is misleading because the accumulated subframes are not lost.

Fixed .

>     * Works in Progressive mode with play enabled.  With play disabled, no effect from changing BufferShader uniforms.  Not sure why yet.

Fixed.

>     * Frags need modifying to put the GUI specification into the BufferShader where it belongs, otherwise they are deleted from the Main VariableEditor because they aren't used in the Main frag, and they aren't included in the BufferShader VariableEditor because they are not defined there.  This probably **breaks loading older saved frags** (which use the hacky workaround putting Post GUI specs in the Main frag even though they are unused there), so this solution may be something to postpone for FragM 2.6 or 3.0?

Fixed partially, only the issue remaining is the presets only working with the Main shader's VariableEditor and not the BufferShader's.

**Comment by claudeha** [comment]
Figured it out: BufferShader's VariableEditor is looking for `#preset`s from the BufferShader frag.  I did some 'orrible 'acking and copied the Main frag presets to the BufferShader presets, but the more I look at this approach (two VariableEditors) the more I dislike it:

- two Preset dropdowns and apply buttons that do the same thing but with the names getting out of sync
- two sets of Reset/Copy/Paste buttons, which only reset/copy/paste the corresponding half of the preset
- inefficient use of screen space (both shaders' VariableEditor are visible)
- implemented with horrible hacks

So I'm open to ideas for better ways to approach this entirely, and won't mind too much if my PR is rejected (there may be some things to cherry-pick into a new attempt, like the parts that move widgets to buffer shader frags).

**Comment by claudeha** [comment]
https://github.com/3Dickulus/FragM/pull/66

**Comment by 3Dickulus** [comment]
encapsulate it in the "Post" tab instead of another variable editor ?

I personally don't mind it the way it was, single variable editor handling all GLSL widgets, the way buffershader widgets are handled is a bit of a trick but works very well

**Comment by claudeha** [comment]
Yes try 2 is a completely different approach that does not do any variable editor shenanigans:

- tracks the "Provenance" of uniforms (the set of shaders they are active in)
- adds any BufferShader widgets to the main VariableEditor
- BufferShader uniforms can have their widgets declared in the Main shader as before; declaring the widgets in the BufferShader works now too
- on widget change, if Provenance is FromBufferShader then don't reset subframes, just render the accumulated back buffer with new settings (allows Post widgets to be really Post rendering)

Main issue remaining is that there is no visual indication of the Provenance of a uniform in the GUI, so you're not sure if you'll lose subframes when touching a widget...

**Comment by 3Dickulus** [comment]
all BufferShader widgets assigned to Post tab, `#group Post` before the uniforms does it. after all Post is intended for BufferShader widgets, no visual indication required re:"Provenance" as adding widgets to the Post tab from the main fragment shader and vice versa doesn't make sense (with the exception of Depth To Alpha). So widgets in the Post tab don't reset subframe count while all others do.

**Comment by claudeha** [comment]
Nothing special about "Post", it is just a tab like any other.  That historically some of the frags put non-post-rendering stuff (like "Gamma", for the non linearGamma path) in the "Post" group is unfortunate, because otherwise it is a fine heuristic.

**Comment by 3Dickulus** [comment]
using DoubleTest...
```
BufferShader-4.frag:uniform double Gamma;
BufferShader-4.frag:            c = dvec3(pow(vec3(c), vec3(2.2))); // It already takes the Gamma into acount
BufferShader-4.frag:    c = pow(vec3(c), vec3(1.0/Gamma));

Progressive2D-4.frag:uniform double Gamma; slider[0.0,2.2,5.0]
Progressive2D-4.frag:#ifdef linearGamma
Progressive2D-4.frag:     FragColor = prev+dvec4( pow( dvec3(new.xyz),dvec3(Gamma)) , new.w);
Progressive2D-4.frag:#ifdef linearGamma
Progressive2D-4.frag:      dvec3 color = pow( vec3(color(c)), vec3(Gamma));
```
is the application of Gamma/linearGamma an accurate description of what is going on mathematically in Progressive2D-4.frag ?
is it the same as BufferShader-4.frag ?
are they (Gamma in both places) tied together ?
can we apply a different name to the variable in Progressive2D-4.frag ?
would this new named variable be appropriately categorized under a different tab ?
can the linearGamma option be translated to a bool switch in the GUI instead of a #define ?
Would it work the opposite way ie: declaring `uniform double Gamma;` in Progressive2D-4.frag and `uniform double Gamma; slider[0.0,2.2,5.0]` in BufferShader-4.frag ? (now that buffershader gui stuff is being addressed properly)

...so many questions :-/

**Comment by claudeha** [comment]
1. I'm not 100% sure what the Gamma stuff is doing there.  All I know is that blending / accumulation should be done in linear light, and I think what the Syntopia frags are doing is expecting nonlinear light from `color()`, applying gamma to make it linear for blending, then undoing the gamma again to make it match the nonlinear light from `color()`.  IMO it would be much saner to just have `color()` do linear light only, and just do Gamma at the end for display on screen (optionally, as EXR is linear for example).  This is what I do in my Trudy frags, I have utilities for converting sRGB to linear that I can use in my user `color()` for converting from uniform values, the buffers are all linear, and the BufferShader converts to sRGB for display.

2. Yes both Gamma values should be the same for accurate results

3. Yes two variables with the same name active in both shaders have only one widget to control them both.  This could cause issues if they have different types, but I haven't added error checking for that case yet.  TODO.

4. You could, but for meaningful results it should take the same value in both shaders. Nonsense can be beautiful sometimes though!

5. You could put Main PreGamma in a Pre tab perhaps, and BufferShader PostGamma in the Post tab.

6. Probably linearGamma could be made a uniform bool, but it changes the meaning of what `color()` outputs, so maybe best to be set by the user frag?  otherwise beautiful nonsense again

7. with these changes, it doesn't matter which shader the widget is defined in (both work, both have the same outcome, except if there are two conflicting widgets for the same uniform in different shaders the one from the Main frag is preferred; no warning message here too, more TODO)

**Comment by 3Dickulus** [comment]
hmm... can the gamma value be a fixed value, like 2.2, between the fragmentshader and the buffershader and use the Gamma slider in Post only for "fine tuning" ? or are we stuck with this just due to the nature of the beast?
If we're stuck with it then so be it, at least one slider in Post tab will reset the subframe count, and we move on.

**Comment by 3Dickulus** [comment]
is this one close-able? or we talkin gamma?
